Solutions/Corvus.Azure.Cosmos.Tenancy/Corvus/Azure/Cosmos/Tenancy/Internal/TenantCosmosContainerFactory.cs

```diff
@@ -154,9 +156,29 @@ public Task<Container> GetContainerForTenantAsync(ITenant tenant, CosmosContaine
             CosmosContainerDefinition tenantedCosmosContainerDefinition = GetContainerDefinitionForTenantAsync(tenant, containerDefinition);
             object key = GetKeyFor(tenantedCosmosContainerDefinition);
 
-            return this.containers.GetOrAdd(
+            Task<Container> result = this.containers.GetOrAdd(
                 key,
-                async _ => await this.CreateTenantCosmosContainer(tenant, containerDefinition, tenantedCosmosContainerDefinition).ConfigureAwait(false));
+                _ => this.CreateTenantCosmosContainer(tenant, containerDefinition, tenantedCosmosContainerDefinition));
+
+            if (result.IsFaulted)
+            {
+                // If a task has been created in the previous statement, it won't have completed yet. Therefore if it's
+                // faulted, that means it was added as part of a previous request to this method, and subsequently
+                // failed. As such, we will remove the item from the dictionary, and attempt to create a new one to
+                // return. If removing the value fails, that's likely because it's been removed by a different thread,
+                // so we will ignore that and just attempt to create and return a new value anyway.
+                this.containers.TryRemove(key, out Task<Container> _);
+
+                // Wait for a short and random time, to reduce the potential for large numbers of spurious container
+                // recreation that could happen if multiple threads are trying to rectify the failure simultanously.
+                await Task.Delay(this.random.Next(150, 250)).ConfigureAwait(false);
+
+                result = this.containers.GetOrAdd(
```

Comment:
What if this is faulted? Do we need to check to see if this is faulted, retry a certain number of times, then actually throw an exception?

Reply:
The exception will be passed on up anyway: the task will be handed back to the caller as soon as it has created, so we'll have at least one caller trying to use this task, and we only retry if there's a further attempt to get the container after the fault has had time to occur.

We could count how many failures we've had but what would we do when hitting a threshold? Decide to stop trying? (That's effectively the policy we have now, albeit with a retry count of 0.)

If there's some transient environmental problem that lasts for a few minutes causing this to fail, then deciding "We've tried too many times, let's give up" might mean that once the underlying problem goes away, recovery now requires manual intervention.

We could report the exception of course, although I think we'd want to report them all instead of waiting for a few to happen first wouldn't we?

Reply:
For me, there's two separate issues here:
1. Fixing the bug that means failure to create the container puts us in an unrecoverable faulted state.
2. Dealing with the failure to create the container.

My feeling is that this PR should deal with item 1, which it now does - it picks up the fact that we've cached a faulted task and removes it from the cache, which gives the consuming code the ability to recover if the failure was transient.

Item 2 - responding to the failure - is, I think, the responsibility of the calling code. It can decide what makes sense depending on the use case. So I'd prefer not to put any retry logic in here and just tell consumers that they need to handle failures and determine their own exception handling/retry policy.